### PR TITLE
[LI-HOTFIX] Stop enqueuing unnecessary LeaderAndIsrResponseReceived events

### PR DESCRIPTION
**Pull request by xiowu0**
TICKET = LIKAFKA-40059
LI_DESCRIPTION =

With li combined request feature, a combined response will cause the controller to enqueue an event of LeaderAndIsrResponseReceived, regardless of whether the combined request contains LeaderAndIsr Request or not.

As a result, a LI combined request with only metadata update would end up generating an empty LeaderAndIsrResponseReceived event in the controller queue when the response is received by the controller.

Since the controller may send out "metadata update requests" to all live brokers, it fills up the controller queue with empty LeaderAndIsrResponseReceived events quickly (e.g., single metadata update request can generate X LeaderAndIsrResponseReceived events in the controller queue, where X = number of live brokers in the cluster).

EXIT_CRITERIA = MANUAL ["when li combined request/response is no longer needed"]

### CODE REVIEW

File: core/src/main/scala/kafka/controller/ControllerRequestMerger.scala [hunk 1]
```diff
@@ -235,10 +236,12 @@ class ControllerRequestMerger extends Logging {
     // Currently, there is no callback for the UpdateMetadataResponse
     val LiDecomposedControlResponse(leaderAndIsrResponse, _, stopReplicaResponse) =
       LiDecomposedControlResponseUtils.decomposeResponse(response.asInstanceOf[LiCombinedControlResponse])
-    if (leaderAndIsrCallback != null) {
+    if (leaderAndIsrCallback != null
+      && (!leaderAndIsrResponse.partitions().isEmpty() || leaderAndIsrResponse.error() != Errors.NONE)) {
```
Comment (efeg) [inline core/src/main/scala/kafka/controller/ControllerRequestMerger.scala]: Nit `isEmpty()` -> `isEmpty`

File: core/src/main/scala/kafka/controller/ControllerRequestMerger.scala [hunk 2]
```diff
@@ -235,10 +236,12 @@ class ControllerRequestMerger extends Logging {
...
       leaderAndIsrCallback(leaderAndIsrResponse)
     }
-    if (stopReplicaCallback != null) {
+    if (stopReplicaCallback != null
+      && (!stopReplicaResponse.partitionErrors().isEmpty()) || stopReplicaResponse.error() != Errors.NONE) {
```
Comment (efeg) [inline core/src/main/scala/kafka/controller/ControllerRequestMerger.scala]: Nit `isEmpty()` -> `isEmpty`

File: core/src/main/scala/kafka/controller/ControllerRequestMerger.scala (hunk 1 above)
Comment (xiowu0) [reply core/src/main/scala/kafka/controller/ControllerRequestMerger.scala]: fixed

**Review by lmr3796 — CHANGES_REQUESTED** [review]
Nit: title seemed too long and may be truncated at merge.

Suggesting this as subject/title and keep the remaining description
```
[LI-HOTFIX] Stop enqueuing unnecessary LeaderAndIsrResponseReceived
```

File: core/src/test/scala/unit/kafka/controller/ControllerRequestMergerTest.scala [hunk 3]
```diff
@@ -273,4 +275,94 @@ class ControllerRequestMergerTest {
       Assert.assertEquals("Mismatched partition states", expectedPartitions(i).asJava, liCombinedControlRequest.stopReplicaPartitionStates())
     }
   }
+
+  @Test
+  def testNotTriggerCallbackForUpdateMetadataReqeust(): Unit = {
```
Comment (efeg) [inline core/src/test/scala/unit/kafka/controller/ControllerRequestMergerTest.scala]: Typo: `Reqeust` -> `Request`

File: core/src/test/scala/unit/kafka/controller/ControllerRequestMergerTest.scala [hunk 4]
```diff
@@ -273,4 +275,94 @@ class ControllerRequestMergerTest {
...
+    // StopReplica response with some errors
+    val responseStopReplicaData1 =
+      new LiCombinedControlResponseData().setStopReplicaErrorCode(Errors.UNKNOWN_SERVER_ERROR.code())
+    val responseStopReplica1 = new LiCombinedControlResponse(responseStopReplicaData1)
+
+    // StopReplica response with non-empty partitions
+    val stopReplicaPartitionsError = createStopReplicaResponsePartitions("foo",
+      Seq(Errors.NONE, Errors.CLUSTER_AUTHORIZATION_FAILED))
+    val responseStopReplicaData2 =
+      new LiCombinedControlResponseData().setStopReplicaPartitionErrors(stopReplicaPartitionsError);
```
Comment (efeg) [inline core/src/test/scala/unit/kafka/controller/ControllerRequestMergerTest.scala]: Nit: semicolon is unnecessary.

File: core/src/test/scala/unit/kafka/controller/ControllerRequestMergerTest.scala [hunk 5]
```diff
@@ -273,4 +275,94 @@ class ControllerRequestMergerTest {
...
+
+    controllerRequestMerger.triggerCallback(responseUpdateMetadata)
+
+    // metadata only response would not trigger any callback
+    Assert.assertEquals("Mismatched callback count", 0, leaderAndISRCallbackInvocationCount)
+    Assert.assertEquals("Mismatched callback count", 0, stopReplicaCallbackInvocationCount)
+
+    // LeaderAndISR response with some errors
+    val responseLeaderAndISRData1 =
+      new LiCombinedControlResponseData().setLeaderAndIsrErrorCode(Errors.UNKNOWN_SERVER_ERROR.code());
```
Comment (efeg) [inline core/src/test/scala/unit/kafka/controller/ControllerRequestMergerTest.scala]: Nit: semicolon is unnecessary.

**Review by efeg — APPROVED** [review]
Noticed a few nits / typos.
Otherwise LGTM.

**Review by lmr3796 — APPROVED** [review]
+1

When doing squash & merge remember to clear the list of commits at the end of description section

**Pull request merged**